File: apps/desktop/src/api/endpoints/convert.ts
import client from './client'

export interface ConvertRequest {
  input_path: string
  output_path: string
  output_format: string
  quality?: string
}

export const convertApi = {
  async pdfToWord(
    inputPath: string,
    outputPath: string,
    quality: string = 'high'
  ): Promise<{ success: boolean; output_path: string }> {
    return client.post('/convert/pdf/to/word', {
      input_path: inputPath,
      output_path: outputPath,
      quality
    })
  },

  async pdfToExcel(
    inputPath: string,
    outputPath: string
  ): Promise<{ success: boolean; output_path: string }> {
    return client.post('/convert/pdf/to/excel', {
      input_path: inputPath,
      output_path: outputPath
    })
  },

  async pdfToPpt(
    inputPath: string,
    outputPath: string
  ): Promise<{ success: boolean; output_path: string }> {
    return client.post('/convert/pdf/to/ppt', {
      input_path: inputPath,
      output_path: outputPath
    })
  },

  async pdfToImage(
    inputPath: string,
    outputDir: string,
    format: string = 'png',
    dpi: number = 150
  ): Promise<{ success: boolean; images: string[] }> {
    return client.post('/convert/pdf/to/image', null, {
      params: {
        input_path: inputPath,
        output_dir: outputDir,
        image_format: format,
        dpi
      }
    })
  },

  async pdfToHtml(
    inputPath: string,
    outputPath: string
  ): Promise<{ success: boolean; output_path: string }> {
    return client.post('/convert/pdf/to/html', {
      input_path: inputPath,
      output_path: outputPath
    })
  },

  async wordToPdf(
    inputPath: string,
    outputPath: string
  ): Promise<{ success: boolean; output_path: string }> {
    return client.post('/convert/word/to/pdf', {
      input_path: inputPath,
      output_path: outputPath
    })
  },

  async excelToPdf(
    inputPath: string,
    outputPath: string
  ): Promise<{ success: boolean; output_path: string }> {
    return client.post('/convert/excel/to/pdf', {
      input_path: inputPath,
      output_path: outputPath
    })
  },

  async imageToPdf(
    inputPaths: string[],
    outputPath: string
  ): Promise<{ success: boolean; output_path: string }> {
    return client.post('/convert/image/to/pdf', {
      input_paths: inputPaths,
      output_path: outputPath
    })
  },

  async htmlToPdf(
    inputPath: string,
    outputPath: string
  ): Promise<{ success: boolean; output_path: string }> {
    return client.post('/convert/html/to/pdf', {
      input_path: inputPath,
      output_path: outputPath
    })
  },

  async getSupportedFormats(): Promise<{
    input_formats: string[]
    output_formats: string[]
  }> {
    return client.get('/convert/formats')
  }
}

export default convertApi

File: apps/server/app/api/v1/convert.py
from fastapi import APIRouter, HTTPException
from typing import Optional
from pydantic import BaseModel

from app.core.convert_engine.engine import ConvertEngine

router = APIRouter()
convert_engine = ConvertEngine()


class ConvertRequest(BaseModel):
    input_path: str
    output_path: str
    output_format: str
    quality: str = "high"


@router.post("/pdf/to/word")
async def pdf_to_word(request: ConvertRequest):
    try:
        result = await convert_engine.pdf_to_word(
            request.input_path,
            request.output_path,
            request.quality
        )
        return {"success": True, "output_path": result}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/pdf/to/excel")
async def pdf_to_excel(request: ConvertRequest):
    try:
        result = await convert_engine.pdf_to_excel(
            request.input_path,
            request.output_path
        )
        return {"success": True, "output_path": result}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/pdf/to/ppt")
async def pdf_to_ppt(request: ConvertRequest):
    try:
        result = await convert_engine.pdf_to_ppt(
            request.input_path,
            request.output_path
        )
        return {"success": True, "output_path": result}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/pdf/to/image")
async def pdf_to_image(
    input_path: str,
    output_dir: str,
    image_format: str = "png",
    dpi: int = 150
):
    try:
        results = await convert_engine.pdf_to_image(
            input_path,
            output_dir,
            image_format,
            dpi
        )
        return {"success": True, "images": results}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/pdf/to/html")
async def pdf_to_html(request: ConvertRequest):
    try:
        result = await convert_engine.pdf_to_html(
            request.input_path,
            request.output_path
        )
        return {"success": True, "output_path": result}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/word/to/pdf")
async def word_to_pdf(request: ConvertRequest):
    try:
        result = await convert_engine.word_to_pdf(
            request.input_path,
            request.output_path
        )
        return {"success": True, "output_path": result}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/excel/to/pdf")
async def excel_to_pdf(request: ConvertRequest):
    try:
        result = await convert_engine.excel_to_pdf(
            request.input_path,
            request.output_path
        )
        return {"success": True, "output_path": result}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/image/to/pdf")
async def image_to_pdf(
    input_paths: list[str],
    output_path: str
):
    try:
        result = await convert_engine.images_to_pdf(input_paths, output_path)
        return {"success": True, "output_path": result}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/html/to/pdf")
async def html_to_pdf(request: ConvertRequest):
    try:
        result = await convert_engine.html_to_pdf(
            request.input_path,
            request.output_path
        )
        return {"success": True, "output_path": result}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/formats")
async def get_supported_formats():
    return {
        "input_formats": ["pdf", "docx", "doc", "xlsx", "xls", "pptx", "ppt", "png", "jpg", "jpeg", "html"],
        "output_formats": ["pdf", "docx", "xlsx", "pptx", "png", "jpg", "html", "txt"]
    }
